### ISSUE TITLE
nasty_meta.p4.test compiler crash

### ISSUE THREAD

**Issue by jrmcclurg** [p4lang/p4c#1924]
Hello, I'm able to `make` p4c just fine, and when I do `make check` all tests pass except for one:

```
p14_to_16/testdata/p4_14_errors/nasty_meta.p4.test -v
```

I get the following error. Any ideas?

```
Writing temporary files into  ./tmpXSMgZi
./p4test /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors/nasty_meta.p4
./p4test --pp ./tmpXSMgZi/nasty_meta.p4 --dump ./tmpXSMgZi --top4 FrontEndLast,FrontEndDump,MidEndLast --testJson --std p4-14 /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors/nasty_meta.p4
Exit code  255
Error compiling
nasty_meta.p4(24): [--Werror=invalid] error: extract: Invalid argument. Expected a header not struct ethernet_t
    extract(ethernet);
    ^^^^^^^^^^^^^^^^^
terminate called after throwing an instance of 'Util::CompilerBug'
  what():  In file: /home/netsynth/tools/p4lang/p4c/lib/crash.cpp:229
Compiler Bug: Exiting with SIGSEGV


Checking nasty_meta.p4-stderr
Comparing /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors_outputs/nasty_meta.p4-stderr and ./tmpXSMgZi/nasty_meta.p4-stderr
diff -B -u -w /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors_outputs/nasty_meta.p4-stderr ./tmpXSMgZi/nasty_meta.p4-stderr >&2
--- /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors_outputs/nasty_meta.p4-stderr	2019-05-06 09:30:37.134037048 -0600
+++ ./tmpXSMgZi/nasty_meta.p4-stderr	2019-05-06 10:01:32.094037048 -0600
@@ -1,4 +1,7 @@
 nasty_meta.p4(24): [--Werror=invalid] error: extract: Invalid argument. Expected a header not struct ethernet_t
     extract(ethernet);
     ^^^^^^^^^^^^^^^^^
-error: 1 errors encountered, aborting compilation
+terminate called after throwing an instance of 'Util::CompilerBug'
+  what():  In file: /home/netsynth/tools/p4lang/p4c/lib/crash.cpp:229
+Compiler Bug: Exiting with SIGSEGV
+
Removing ./tmpXSMgZi
Program was expected to fail
```

**Comment by ChrisDodd** [comment]
Try running the test with `-Tcrash:1` to get a crash dump?

**Comment by jrmcclurg** [comment]
Here's the output with `-Tcrash:1`. BTW I found that if I go back to a [Mar 16 commit](https://github.com/p4lang/p4c/commit/43fe4bf1668802a7e9bedd29be2f82fe1b69cbc1), I can `make check` properly.

```
Writing temporary files into  ./tmpSzIV66
./p4test /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors/nasty_meta.p4
./p4test --pp ./tmpSzIV66/nasty_meta.p4 --dump ./tmpSzIV66 --top4 FrontEndLast,FrontEndDump,MidEndLast --testJson -Tcrash:1 --std p4-14 /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors/nasty_meta.p4
Exit code  255
Error compiling
nasty_meta.p4(24): [--Werror=invalid] error: extract: Invalid argument. Expected a header not struct ethernet_t
    extract(ethernet);
    ^^^^^^^^^^^^^^^^^
exiting with SIGSEGV, trace:
  address = 0
 rax=0000000000000000 rbx=0000000000000000 rcx=fffffffffffffffe
 rdx=000055e5d04cae60 rdi=0000000000000000 rsi=000055e5d04d2748
 rbp=0000000000000000 rsp=00007ffe55461420  r8=0000000000000001
  r9=0000000000000000 r10=000000000000000f r11=00007f2790a3b9f0
 r12=0000000000000000 r13=00007ffe55461680 r14=00007ffe554616e0
 r15=00007ffe55461820
  /lib/x86_64-linux-gnu/libc.so.6(+0x3ef20) [0x7f278f5b3f20]
  /usr/lib/x86_64-linux-gnu/libstdc++.so.6(__dynamic_cast+0x1d) [0x7f278fc0f3fd]
  ./p4test(_ZN2IR13IndexedVectorINS_10StatOrDeclEE11insertInMapEPKS1_+0x3f) [0x55e5cf99d41f]
  ./p4test(_ZN4P4V116ProgramStructure13convertParserEPKN2IR8V1ParserEPNS1_13IndexedVectorINS1_11DeclarationEEE+0xc5) [0x55e5cfbc7905]
  ./p4test(_ZN4P4V116ProgramStructure12createParserEv+0x54a) [0x55e5cfbdbb4a]
  ./p4test(_ZN4P4V116ProgramStructure6createEN4Util10SourceInfoE+0x49) [0x55e5cfbc5c89]
  ./p4test(+0x115bf95) [0x55e5cfbc5f95]
  ./p4test(_ZN9Transform13apply_visitorEPKN2IR4NodeEPKc+0x1ef) [0x55e5cfda541f]
  ./p4test(_ZNK2IR4Node5applyER7Visitor+0x3b) [0x55e5cfd6f85b]
  ./p4test(_ZN11PassManager13apply_visitorEPKN2IR4NodeEPKc+0x172) [0x55e5cfda3e72]
  ./p4test(_ZNK2IR4Node5applyER7Visitor+0x3b) [0x55e5cfd6f85b]
  ./p4test(_ZN2P411parseP4FileER15CompilerOptions+0x1f7) [0x55e5cfb6cf87]
  ./p4test(main+0x1ce) [0x55e5cf88bafe]
  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f278f596b97]
  ./p4test(_start+0x2a) [0x55e5cf8928ca]
terminate called after throwing an instance of 'Util::CompilerBug'
  what():  In file: /home/netsynth/tools/p4lang/p4c/lib/crash.cpp:229
Compiler Bug: Exiting with SIGSEGV


Checking nasty_meta.p4-stderr
Comparing /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors_outputs/nasty_meta.p4-stderr and ./tmpSzIV66/nasty_meta.p4-stderr
diff -B -u -w /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors_outputs/nasty_meta.p4-stderr ./tmpSzIV66/nasty_meta.p4-stderr >&2
--- /home/netsynth/tools/p4lang/p4c/testdata/p4_14_errors_outputs/nasty_meta.p4-stderr	2019-05-06 10:34:12.846037048 -0600
+++ ./tmpSzIV66/nasty_meta.p4-stderr	2019-05-06 10:39:06.346037048 -0600
@@ -1,4 +1,30 @@
 nasty_meta.p4(24): [--Werror=invalid] error: extract: Invalid argument. Expected a header not struct ethernet_t
     extract(ethernet);
     ^^^^^^^^^^^^^^^^^
-error: 1 errors encountered, aborting compilation
+exiting with SIGSEGV, trace:
+  address = 0
+ rax=0000000000000000 rbx=0000000000000000 rcx=fffffffffffffffe
+ rdx=000055e5d04cae60 rdi=0000000000000000 rsi=000055e5d04d2748
+ rbp=0000000000000000 rsp=00007ffe55461420  r8=0000000000000001
+  r9=0000000000000000 r10=000000000000000f r11=00007f2790a3b9f0
+ r12=0000000000000000 r13=00007ffe55461680 r14=00007ffe554616e0
+ r15=00007ffe55461820
+  /lib/x86_64-linux-gnu/libc.so.6(+0x3ef20) [0x7f278f5b3f20]
+  /usr/lib/x86_64-linux-gnu/libstdc++.so.6(__dynamic_cast+0x1d) [0x7f278fc0f3fd]
+  ./p4test(_ZN2IR13IndexedVectorINS_10StatOrDeclEE11insertInMapEPKS1_+0x3f) [0x55e5cf99d41f]
+  ./p4test(_ZN4P4V116ProgramStructure13convertParserEPKN2IR8V1ParserEPNS1_13IndexedVectorINS1_11DeclarationEEE+0xc5) [0x55e5cfbc7905]
+  ./p4test(_ZN4P4V116ProgramStructure12createParserEv+0x54a) [0x55e5cfbdbb4a]
+  ./p4test(_ZN4P4V116ProgramStructure6createEN4Util10SourceInfoE+0x49) [0x55e5cfbc5c89]
+  ./p4test(+0x115bf95) [0x55e5cfbc5f95]
+  ./p4test(_ZN9Transform13apply_visitorEPKN2IR4NodeEPKc+0x1ef) [0x55e5cfda541f]
+  ./p4test(_ZNK2IR4Node5applyER7Visitor+0x3b) [0x55e5cfd6f85b]
+  ./p4test(_ZN11PassManager13apply_visitorEPKN2IR4NodeEPKc+0x172) [0x55e5cfda3e72]
+  ./p4test(_ZNK2IR4Node5applyER7Visitor+0x3b) [0x55e5cfd6f85b]
+  ./p4test(_ZN2P411parseP4FileER15CompilerOptions+0x1f7) [0x55e5cfb6cf87]
+  ./p4test(main+0x1ce) [0x55e5cf88bafe]
+  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f278f596b97]
+  ./p4test(_start+0x2a) [0x55e5cf8928ca]
+terminate called after throwing an instance of 'Util::CompilerBug'
+  what():  In file: /home/netsynth/tools/p4lang/p4c/lib/crash.cpp:229
+Compiler Bug: Exiting with SIGSEGV
+
Removing ./tmpSzIV66
Program was expected to fail
```

**Comment by antoninbas** [comment]
I'm glad someone is seeing this issue. The same test has been failing on launchpad and I have been unable to reproduce this locally: https://launchpadlibrarian.net/422444694/buildlog_ubuntu-bionic-amd64.p4c_1.1.0-rc1-201904302116-60be53f~ubuntu18.04.1_BUILDING.txt.gz.
@frederic-loui this may interest you

**Comment by jrmcclurg** [comment]
Just to give a bit more background:
```
# uname -a
Linux netsynth-demo 4.18.0-18-generic #19~18.04.1-Ubuntu SMP Fri Apr 5 10:22:13 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename:	bionic
```

I'm also using Protobuf v3.3.0 because the version v3.2.0 in the instructions has some build errors.

**Comment by ChrisDodd** [comment]
That's a weird trace -- the crash indicates a nullptr in dynamic_cast, but dynamic_cast<T *>(nullptr) should always be safe (and return nullptr), so I'm not sure why it is crashing.

**Comment by jrmcclurg** [comment]
I was able to confirm that [this Mar 25 commit](https://github.com/p4lang/p4c/commit/f7035240a0ec91cfae7389e1ab6ca7f251f57418) is what introduces the problem on my machine.

**Comment by cc10512** [comment]
That's expected, as that is the commit that introduced the test. I tried to reproduce it before on an Ubuntu VM and couldn't. What compiler version are you using?

**Comment by jrmcclurg** [comment]
I came up with a fix that works on my machine:

in `ir/indexed_vector.h`, I added the following null-check in `insertInMap`
````
void insertInMap(const T* a) {
        if (a == nullptr || !a->template is<IDeclaration>())
```

There might be a similar problem in `removeFromMap`

**Comment by jrmcclurg** [comment]
By compiler version, do you mean g++?
```
g++ (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0
```

**Comment by cc10512** [comment]
Yes, your fix is a good defense. 

Can you also please try the following and see if it fixes the issue?
```
diff --git a/frontends/p4/fromv1.0/programStructure.cpp b/frontends/p4/fromv1.0/programStructure.cpp
index 5027f351..c88b35a0 100644
--- a/frontends/p4/fromv1.0/programStructure.cpp
+++ b/frontends/p4/fromv1.0/programStructure.cpp
@@ -421,7 +421,7 @@ const IR::ParserState* ProgramStructure::convertParser(const IR::V1Parser* parse
     IR::IndexedVector<IR::StatOrDecl> components;
     for (auto e : parser->stmts) {
         auto stmt = convertParserStatement(e);
-        components.push_back(stmt);
+        if (stmt) components.push_back(stmt);
     }
     const IR::Expression* select = nullptr;
     if (parser->select != nullptr) {
```

**Comment by jrmcclurg** [comment]
@cc10512 yes, that works! All tests pass on my machine.

Are you able to push this fix to master?

**Comment by frederic-loui** [comment]
Hi, I initially planned to investigate the problem observed systematically on launchpad as it might be linked to the packaging set. But not sure why, it disappeared when initiating `make check` interactively and from time to time appeared again. As a side note, I also removed optimisation flag (-O3 used in p4c Dockerfile) which appear to lead to tests failure. (I compiled the same code with and without the flag, twice in a row and some tests in error disappeared, except this nasty_meta.p4) I'm glad you nailed this issue @jrmcclurg. As soon as the patch goes to master we should see the first nightly p4c bionic package. (thanks @cc10512 ) :-)
